### PR TITLE
Add explicit check for number of channels

### PR DESCRIPTION
**Pull request by ademyanchuk**
Example why you need to check it:
`M = torch.randint(low=0, high=2, size=(6, 64, 64), dtype = torch.float)`
When you put this input through to_pil_image without mode argument, it converts to uint8 here:
```
if pic.is_floating_point() and mode != 'F':
            pic = pic.mul(255).byte()
```
and change the mode to RGB here:
```
if mode is None and npimg.dtype == np.uint8:
            mode = 'RGB'
```
Image.fromarray doesn't raise if provided with mode RGB and just cut number of channels from what you have to 3

### CODE REVIEW

**Comment by ademyanchuk** [comment]
Also in this Pytorch forum [discussion](https://discuss.pytorch.org/t/transformation-is-reducing-the-number-of-channel/103033).

**Comment by vfdev-5** [comment]
@ademyanchuk thanks for the PR ! I'd add the check before to numpy conversion, before line 194, as `if pic.shape[-3] > 4`.
It would be good to add a test on that too, for example here:
https://github.com/pytorch/vision/blob/aa4cf03961f93fd8290b8a57ca1052d01dfc0f5c/test/test_transforms.py#L989
and 
https://github.com/pytorch/vision/blob/aa4cf03961f93fd8290b8a57ca1052d01dfc0f5c/test/test_transforms.py#L993
Thanks

**Comment by ademyanchuk** [comment]
@vfdev-5 thanks for comment! I thought where to put the check for number of channels and decided to put it after conversion to numpy as at this point we know the channel dimension position is `.shape[2]`. Given the function takes as input both `torch.tensor` and `numpy.array` and assumes `tensor` is channel first, we have to add check two times for `numpy` with `ndim==3` and for `torch` with `ndim==3`. Do you think I should add the check two times? Or perhaps is it better to check only once after conversion?

Sure, I would like to add tests. I will see how can I add those. It is my first PR :)

**Comment by vfdev-5** [comment]
> Given the function takes as input both torch.tensor and numpy.array and assumes tensor is channel first, we have to add check two times for numpy with ndim==3 and for torch with ndim==3.

Yes, true. For tensor the assumption is that number of channels is `-3` and for numpy is `-1`. Idea is that IMO it would be better to raise exception before making data processing ...

> Sure, I would like to add tests. I will see how can I add those. It is my first PR :)

Sounds good. Feel free to ask questions if needed. Please, take a look at drafted [CONTRIBUTING](https://github.com/pytorch/vision/pull/2663), if needed.

**Comment by ademyanchuk** [comment]
> Idea is that IMO it would be better to raise exception before making data processing ...

This is totally make sense. So I will proceed.

**Comment by ademyanchuk** [comment]
Ok. I refactored the checking to be done before processing and added a test for invalid number of channels :)

File: test/test_transforms.py [hunk 1]
```diff
@@ -989,6 +989,7 @@ def test_2d_ndarray_to_pil_image(self):
     def test_tensor_bad_types_to_pil_image(self):
         with self.assertRaises(ValueError):
             transforms.ToPILImage()(torch.ones(1, 3, 4, 4))
+            transforms.ToPILImage()(torch.ones(6, 4, 4))
```
Comment (vfdev-5) [inline test/test_transforms.py:992]: This won't work, actually. You can try locally to replace your line by `assert False` and test still passes.
You have to do
```python
with self.assertRaises(ValueError):
    transforms.ToPILImage()(torch.ones(1, 3, 4, 4))

with self.assertRaises(ValueError):
    transforms.ToPILImage()(torch.ones(6, 4, 4))
```

Let's also improve the test codebase, by replacing `self.assertRaises(ValueError)` by `self.assertRaisesRegex(ValueError, r"REGEX MESSAGE TO CATCH"):`

File: test/test_transforms.py [hunk 2]
```diff
@@ -1000,6 +1001,7 @@ def test_ndarray_bad_types_to_pil_image(self):
 
         with self.assertRaises(ValueError):
             transforms.ToPILImage()(np.ones([1, 4, 4, 3]))
+            transforms.ToPILImage()(np.ones([4, 4, 6]))
```
Comment (vfdev-5) [inline test/test_transforms.py:1004]: Same here

File: torchvision/transforms/functional.py [hunk 3]
```diff
@@ -179,6 +179,11 @@ def to_pil_image(pic, mode=None):
         if pic.ndimension() not in {2, 3}:
             raise ValueError('pic should be 2/3 dimensional. Got {} dimensions.'.format(pic.ndimension()))
 
+        elif pic.ndimension() == 3:
```
Comment (vfdev-5) [inline torchvision/transforms/functional.py:182]: Below we use unsqueeze to make pic as 3d tensor, so let's check and raise exception after the unsqueeze (which is not that expensive). Please, use also `pic.shape[-3]` instead of `pic.shape[0]` and no need to check `elif pic.ndimension() == 3`.

File: torchvision/transforms/functional.py [hunk 4]
```diff
@@ -187,6 +192,11 @@ def to_pil_image(pic, mode=None):
         if pic.ndim not in {2, 3}:
             raise ValueError('pic should be 2/3 dimensional. Got {} dimensions.'.format(pic.ndim))
 
+        elif pic.ndim == 3:
+            # check number of channels
+            if pic.shape[2] > 4:
```
Comment (vfdev-5) [inline torchvision/transforms/functional.py:197]: Same here and please use `pic.shape[-1]` instead of `pic.shape[2]`.

**Review by vfdev-5 — COMMENTED** [review]
Thanks for the updates @ademyanchuk ! I left some comments to improve the PR.

File: torchvision/transforms/functional.py (hunk 4 above)
Comment (ademyanchuk) [reply torchvision/transforms/functional.py:197]: Is it some convention to use minus-indexing? Does it have some practicality? I am curious

Comment (vfdev-5) [reply torchvision/transforms/functional.py:197]: Well, it is a sort of convention that the channel is the last index for numpy and 3rd from right for tensors. For this method it does not matter as input tensor or numpy array should be 2 or 3 dims but for other methods accepting tensors with batch dimension, index=0 points to batch dim instead of channel dim.

Comment (ademyanchuk) [reply torchvision/transforms/functional.py:197]: Ok, now I see the point it's practical to count from the right and also for consistency perhaps

File: test/test_transforms.py (hunk 1 above)
Comment (ademyanchuk) [reply test/test_transforms.py:992]: For `self.assertRaisesRegex(ValueError, r"REGEX MESSAGE TO CATCH"):`
should i check for exact number or could I just check it like this?
`with self.assertRaisesRegex(ValueError, r'pic should not have > 4 channels.'):`

Comment (ademyanchuk) [reply test/test_transforms.py:992]: Also should I update all the code in these functions to use `self.assertRaisesRegex`?

Comment (vfdev-5) [reply test/test_transforms.py:992]: > Also should I update all the code in these functions to use self.assertRaisesRegex?

yes, please.

You can do `self.assertRaisesRegex(ValueError, r'pic should not have > \.+ channels.')` or similar as it is an regular expression to match a message.

**Comment by ademyanchuk** [comment]
Ok. It seems like I manage to make all discussed changes :)

File: test/test_transforms.py [hunk 5]
```diff
@@ -987,19 +987,23 @@ def test_2d_ndarray_to_pil_image(self):
...
-        with self.assertRaises(ValueError):
+        with self.assertRaisesRegex(ValueError, r'pic should be 2/3 dimensional. Got \d+ dimensions.'):
             transforms.ToPILImage()(torch.ones(1, 3, 4, 4))
+        with self.assertRaisesRegex(ValueError, r'pic should not have > 4 channels. Got \d+ channels.'):
+            transforms.ToPILImage()(torch.ones(6, 4, 4))
 
     def test_ndarray_bad_types_to_pil_image(self):
         trans = transforms.ToPILImage()
-        with self.assertRaises(TypeError):
+        with self.assertRaisesRegex(TypeError, r'Input type \w+ is not supported'):
```
Comment (vfdev-5) [inline test/test_transforms.py:997]: Could you please also fix this test as it was also incorrect: add `with self.assertRaisesRegex(TypeError, r'Input type \w+ is not supported')` for each `trans(np.ones([4, 4, 1], type))`.

File: torchvision/transforms/functional.py (hunk 3 above)
Comment (vfdev-5) [reply torchvision/transforms/functional.py:182]: This code can be removed ?

File: torchvision/transforms/functional.py (hunk 4 above)
Comment (vfdev-5) [reply torchvision/transforms/functional.py:201]: This code can be removed, right ?

**Review by vfdev-5 — COMMENTED** [review]
Thanks for the updates @ademyanchuk 
I left few other comments

**Comment by ademyanchuk** [comment]
@vfdev-5 , sorry to bather. How could I do `python setup.py develop` so it compiles with cuda version 10.2?

**Comment by vfdev-5** [comment]
> How could I do python setup.py develop so it compiles with cuda version 10.2?

It will build cuda extensions with the same cuda version as torch has. If you'd like to disable it, just set 
```
CUDA_VISIBLE_DEVICES="" python setup.py develop
```

**Comment by ademyanchuk** [comment]
> It will build cuda extensions with the same cuda version as torch has

Apparently in my case it is not :( 
I did `conda install pytorch -c pytorch-nightly`
it install torch with cuda 10.2
but `python setup.py develop` compile cuda11 version

**Comment by ademyanchuk** [comment]
Anyway, I made the changes from your last comments :)

**Comment by vfdev-5** [comment]
> > It will build cuda extensions with the same cuda version as torch has
> 
> Apparently in my case it is not :(
> I did `conda install pytorch -c pytorch-nightly`
> it install torch with cuda 10.2
> but `python setup.py develop` compile cuda11 version
> 
> On my local machine cuda 11 is installed - could it be issue?
> 
> So, as it is WSL2, i just delete my local cuda11. And torchvision compiled with matched version of cuda as in torch nightly!

It would be nice to have a detailed instructions to be able to reproduce your situation. If you could reproduce the issue, please file an issue about that.

**Comment by vfdev-5** [comment]
Could you also please merge master to your branch. There is unexpected modification here : https://github.com/pytorch/vision/pull/3013/files#diff-b5826ce59a98ee20294c65d06e09363609879221494b5ff7fbb7087b75f3e159L7

**Comment by ademyanchuk** [comment]
Hi, I did merge master into my branch patch-1. I also manually checked this file. But I have no diff with pytorch vision master. This diff that you show the link for is from this commit (as I understand) https://github.com/pytorch/vision/pull/3020/commits/365f1594fc12937e31b0844261d95ca31025c1dd

**Comment by ademyanchuk** [comment]
> It would be nice to have a detailed instructions to be able to reproduce your situation. If you could reproduce the issue, please file an issue about that.

I will try to reproduce the issue. If I succeed, I would file the issue.

**Review by vfdev-5 — APPROVED** [review]
Thanks @ademyanchuk !

**Comment by ademyanchuk** [comment]
Thanks @vfdev-5 for a valuable comments and review. Appreciate! So, is it done? How the merge in master is proceeded?

**Comment by ademyanchuk** [comment]
It appears to me like most of recent pull requests fail on the same travis ci job on the same place.

**Review by fmassa — APPROVED** [review]
